# ~

## Set-up

### Imports

In [1]:
# Imports - standard
import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
# Imports - custom
import sys
sys.path.append(f"../code")
from paths import PROJECT_PATH, DATASET_PATH
from info import PATIENTS, FS, MATERIALS
from settings import *

In [3]:
# auto reload
%load_ext autoreload
%autoreload 2

### Settings

In [4]:
# 

### Functions

## Main

### Task-modulated electrodes

In [9]:
# load results of step 3 and merge with electrode info
fname = f"{PROJECT_PATH}/data/results/band_power_statistics.csv"
temp = pd.read_csv(fname, index_col=0)
temp = temp.loc[temp['memory']=='hit']
df_w = temp.loc[temp['material']=='words']
df_f = temp.loc[temp['material']=='faces']
for df in [df_w, df_f]: # compute joint significance
    df['sig_all'] = df[[f'{band}_sig' for band in BANDS]].all(axis=1)
    df['sig_any'] = df[[f'{band}_sig' for band in BANDS]].any(axis=1)


C:\Users\micha\AppData\Local\Temp\ipykernel_119884\3631205475.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['sig_all'] = df[[f'{band}_sig' for band in BANDS]].all(axis=1)
C:\Users\micha\AppData\Local\Temp\ipykernel_119884\3631205475.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['sig_any'] = df[[f'{band}_sig' for band in BANDS]].any(axis=1)
C:\Users\micha\AppData\Local\Temp\ipykernel_119884\3631205475.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a Dat

(104, 97)

In [15]:
# print number (and percent) of significant channels for each band and both bands for each material

for df, material in zip([df_w, df_f], MATERIALS):
    print(material)
    for band in BANDS:
        print(f"{band}: {df[band+'_sig'].sum()} ({df[band+'_sig'].mean()*100:.1f}%)")
    print(f"joint: {df['sig_all'].sum()} ({df['sig_all'].mean()*100:.1f}%)")
    print()

words
alpha: 248 (35.7%)
gamma: 238 (34.2%)
joint: 104 (15.0%)

faces
alpha: 235 (33.8%)
gamma: 221 (31.8%)
joint: 97 (14.0%)



### SpecParam r-squared

In [17]:
from specparam import SpectralGroupModel
from specparam_utils import compute_adj_r2


In [18]:
r2_knee = []
r2_fixed = []

# loop through conditions
for material in ['words','faces']:
    for memory in ['hit','miss']:
        for window in ['pre', 'post']:
            # load r-squared values for 'knee' model
            fg_k = SpectralGroupModel()
            fname = f"psd_{material}_{memory}_{window}stim_params_knee"
            fg_k.load(f'{PROJECT_PATH}/data/ieeg_psd_param/{fname}')
            r2_k = compute_adj_r2(fg_k)
            r2_knee.append(r2_k)

            # load r-squared values for 'fixed' model
            fg_f = SpectralGroupModel()
            fname = f"psd_{material}_{memory}_{window}stim_params_fixed"
            fg_f.load(f'{PROJECT_PATH}/data/ieeg_psd_param/{fname}')
            r2_f = compute_adj_r2(fg_f)
            r2_fixed.append(r2_f)



In [22]:
print(f"Fixed model: {np.nanmean(r2_fixed):.3f}")
print(f"Knee model: {np.nanmean(r2_knee):.3f}")

Fixed model: 0.960
Knee model: 0.976


### Hierarchical bootstrap

In [23]:
fname = r"C:\Users\micha\projects\oscillation_vs_exponent\data\ieeg_stats\group_level_hierarchical_bootstrap_active_bymaterial.csv"
df = pd.read_csv(fname, index_col=0)
df = df.loc[df['memory']=='hit']
df

material memory    feature  pvalue  cohens_d
0     faces    hit   exponent   0.000  0.439738
1     faces    hit      alpha   0.006  0.173087
2     faces    hit  alpha_adj   0.040  0.207368
3     faces    hit      gamma   0.003 -0.141586
4     faces    hit  gamma_adj   0.273 -0.131028
10    words    hit   exponent   0.000  0.358535
11    words    hit      alpha   0.004  0.143657
12    words    hit  alpha_adj   0.002  0.255376
13    words    hit      gamma   0.003 -0.089741
14    words    hit  gamma_adj   0.146 -0.285829

In [25]:
# correct for multiple comparisons
from statsmodels.stats.multitest import multipletests
pvalues = df['pvalue']
pvalues[pvalues==0] = 0.001
stats = multipletests(pvalues, alpha=0.05, method='holm')
df['p_corr'] = stats[1]
df['sig'] = df['p_corr'] < 0.05
df

material memory    feature  pvalue  cohens_d  p_corr    sig
0     faces    hit   exponent   0.001  0.439738   0.010   True
1     faces    hit      alpha   0.006  0.173087   0.024   True
2     faces    hit  alpha_adj   0.040  0.207368   0.120  False
3     faces    hit      gamma   0.003 -0.141586   0.021   True
4     faces    hit  gamma_adj   0.273 -0.131028   0.292  False
10    words    hit   exponent   0.001  0.358535   0.010   True
11    words    hit      alpha   0.004  0.143657   0.021   True
12    words    hit  alpha_adj   0.002  0.255376   0.016   True
13    words    hit      gamma   0.003 -0.089741   0.021   True
14    words    hit  gamma_adj   0.146 -0.285829   0.292  False

### Intersection Frequency

In [37]:
data_in = np.load(f"{PROJECT_PATH}/data/ieeg_intersection_results/intersection_results_faces_hit_knee.npz")
intersection_f = data_in['intersection'][df_f['sig_all'].values]
data_in = np.load(f"{PROJECT_PATH}/data/ieeg_intersection_results/intersection_results_words_hit_knee.npz")
intersection_w = data_in['intersection'][df_w['sig_all'].values]

# print median and std
print(f"Median intersection:")
print(f"    Words:\n\tmedian: {np.nanmedian(intersection_w):.3f}\n\tstd: {np.nanstd(intersection_w):.3f}")
print(f"    Faces:\n\tmedian: {np.nanmedian(intersection_f):.3f}\n\tstd: {np.nanstd(intersection_f):.3f}")

Median intersection:
    Words:
	median: 32.936
	std: 20.863
    Faces:
	median: 37.926
	std: 20.803


### Regression of confounding variables

---------------------------------------  
Word-encoding  
---------------------------------------  
alpha results:  
        R-squared: 0.177  
        F-statistic: 287.733  
        p-value: 1.356e-58  

alpha_adj results:  
        R-squared: 0.003  
        F-statistic: 4.434  
        p-value: 0.035  

gamma results:  
        R-squared: 0.229  
        F-statistic: 396.464  
        p-value: 1.887e-77

gamma_adj results:  
        R-squared: 0.001  
        F-statistic: 1.531  
        p-value: 0.216  


alpha cross-validation t-test results:  
        T-statistic: 4.013  
        p-value: 0.002  

gamma cross-validation t-test results:  
        T-statistic: 3.315  
        p-value: 0.006  

---------------------------------------  
Face-encoding  
---------------------------------------    
alpha results:  
        R-squared: 0.131  
        F-statistic: 99.716  
        p-value: 5.716e-22  

alpha_adj results:  
        R-squared: 0.001  
        F-statistic: 0.656  
        p-value: 0.418  

gamma results:  
        R-squared: 0.174  
        F-statistic: 139.814  
        p-value: 2.087e-29  

gamma_adj results:  
        R-squared: 0.005  
        F-statistic: 3.403  
        p-value: 0.066  

alpha cross-validation t-test results:  
        T-statistic: 3.405  
        p-value: 0.005  

gamma cross-validation t-test results:  
        T-statistic: 1.788  
        p-value: 0.099  


### Behavioral performance

In [40]:
from mne import read_epochs

# load behavioral data for all patients
dir_input = f"{PROJECT_PATH}/data/ieeg_epochs/"
files = os.listdir(dir_input)
df_list = []
for fname in files:
    f_parts = fname.split('_')
    epochs = read_epochs(f"{dir_input}/{fname}", verbose=False)
    df_i = epochs.metadata
    df_i.insert(0, 'patient', f_parts[0])
    df_i.insert(1, 'material', f_parts[1])
    df_list.append(df_i)
df = pd.concat(df_list)
df

patient material  trial_num  pleasantness  confidence  recalled  \
0    pat02    faces        2.0           4.0         3.0       1.0   
2    pat02    faces        4.0           2.0         1.0       1.0   
3    pat02    faces        5.0           2.0         2.0       1.0   
4    pat02    faces        6.0           1.0         2.0       1.0   
5    pat02    faces        7.0           3.0         1.0       1.0   
..     ...      ...        ...           ...         ...       ...   
83   pat22    words       86.0           3.0         4.0       0.0   
85   pat22    words       88.0           3.0         4.0       0.0   
92   pat22    words       95.0           5.0         5.0       0.0   
93   pat22    words       96.0           5.0         4.0       0.0   
96   pat22    words       99.0           4.0         4.0       0.0   

    reaction_time  
0          1780.1  
2          2208.4  
3          2369.6  
4          2301.8  
5          2365.2  
..            ...  
83         4964.1  
85         5109.9  
92         3632.1  
93         4930.6  
96         4499.8  

[2583 rows x 7 columns]

In [41]:
# count number of trials and hits per patient
behav_list = [] 
patients = df['patient'].unique()
for material in ['faces', 'words']:
    df_i = df.loc[df['material']==material]
    n_trials = df_i.groupby('patient').size().values
    n_correct = df_i.groupby('patient').sum()['recalled'].values.astype(int)
    behav_i = pd.DataFrame({'patient': patients, 'material': material,
        'n_trials': n_trials, 'n_correct': n_correct})
    behav_list.append(behav_i)
behav = pd.concat(behav_list)
behav.insert(4, 'pct_correct', behav['n_correct']/behav['n_trials'])
behav

patient material  n_trials  n_correct  pct_correct
0    pat02    faces        99         63     0.636364
1    pat04    faces        99         60     0.606061
2    pat05    faces        97         65     0.670103
3    pat08    faces        94         56     0.595745
4    pat10    faces       100         66     0.660000
5    pat11    faces       100         44     0.440000
6    pat15    faces       100         49     0.490000
7    pat16    faces       100         21     0.210000
8    pat17    faces       100         35     0.350000
9    pat19    faces       100         51     0.510000
10   pat20    faces       100         41     0.410000
11   pat21    faces       100         73     0.730000
12   pat22    faces       100         58     0.580000
0    pat02    words        99         78     0.787879
1    pat04    words        99         54     0.545455
2    pat05    words        97         80     0.824742
3    pat08    words        99         76     0.767677
4    pat10    words       100         72     0.720000
5    pat11    words       100         55     0.550000
6    pat15    words       100         56     0.560000
7    pat16    words       100         67     0.670000
8    pat17    words       100         74     0.740000
9    pat19    words       100         85     0.850000
10   pat20    words       100         70     0.700000
11   pat21    words       100         77     0.770000
12   pat22    words       100         58     0.580000

In [45]:
# print mean accuracy for each block
print(f"Overall mean: {behav['pct_correct'].mean()}")
print(behav.iloc[:,1:].groupby('material').mean()['pct_correct'])

Overall mean: 0.6136163225042365
material
faces    0.529867
words    0.697366
Name: pct_correct, dtype: float64
